### ISSUE TITLE
Error parsing Timezone names with "+"

### ISSUE THREAD

**Issue by marcelog** [bitwalker/timex#203]
Hello,

This is related to #199.

The character `+` is missing from the list of allowed characters for a time zone name, so functions like `parse!` fail with valid timezones like `Etc/GMT+0` (with _valid_ in this case I'm referring to time zone names that appear in the list returned by `Timex.timezones/0`).

Here's a sample run:

``` elixir
Timex.parse! "2016-01-01T19:32:32 Etc/GMT+0", "%Y-%m-%dT%H:%M:%S %Z", :strftime
** (Timex.Parse.ParseError) Expected end of input at line 1, column 27
    (timex) lib/parse/datetime/parser.ex:96: Timex.Parse.DateTime.Parser.parse!/3
```

The character `+` can be added in https://github.com/bitwalker/timex/blob/master/lib/parse/datetime/parsers.ex#L157 with:

``` elixir
    word_of(~r/[\/\w_+-]/)
```

Although this makes this test fail (https://github.com/bitwalker/timex/blob/master/test/parse_strftime_test.exs#L12):

``` elixir
assert {:ok, ^date} = parse("Mon Jul 06 2015 00:00:00 GMT+0200 (CST)", "%a %b %d %Y %H:%M:%S %Z%z (%Z)")
```

Because now the offset is getting swallowed as part of the time zone name. Due to time constraints I haven't worked on a full patch, and I also wanted to know your thoughts first about this issue.

Thanks in advance!

**Comment by bitwalker** [comment]
There is two ways to look at this. One the one hand, `%Z%z` should give you the same result as `%Z` with the fix you mentioned, as `GMT` is the base timezone name, and `+0200` is the offset, but due to the semantics of `GMT+0200` (it's actually `-2` hours from GMT, rather than `+2`), that likely would give an incorrect result. On the other hand, my feeling is that `%Z` should be able to parse `GMT+0200`, and in this case it's fine that the offset is swallowed, since it's technically part of the timezone name. Which means that particular test should simply use `%Z` instead of `%Z%z`.

Do you happen to know offhand how other strftime implementations handle this?

**Comment by marcelog** [comment]
I see your point.. in Ruby, it seems like indeed %Z swallows the timezone name:

```ruby
> DateTime.strptime 'Mon Jul 06 2015 00:00:00 GMT+0200 (CST)', '%a %b %d %Y %H:%M:%S %Z%z (%Z)'
ArgumentError: invalid date
```

Adding the offset overrides the base time zone:
```ruby
> DateTime.strptime 'Mon Jul 06 2015 00:00:00 GMT+0200+0300', '%a %b %d %Y %H:%M:%S %Z%z'
 => #<DateTime: 2015-07-06T00:00:00+03:00 ((2457209j,75600s,0n),+10800s,2299161j)>
```

Using another time zone as base:
```ruby
> DateTime.strptime 'Mon Jul 06 2015 00:00:00 CST+0200', '%a %b %d %Y %H:%M:%S %Z%z'
 => #<DateTime: 2015-07-06T00:00:00+02:00 ((2457209j,79200s,0n),+7200s,2299161j)>
```

And it seems that in this case, both `%z` and `%Z` are accepted:

```ruby
> DateTime.strptime 'Mon Jul 06 2015 00:00:00 GMT+0200', '%a %b %d %Y %H:%M:%S %z'
 => #<DateTime: 2015-07-06T00:00:00+02:00 ((2457209j,79200s,0n),+7200s,2299161j)>
```

```ruby
> DateTime.strptime 'Mon Jul 06 2015 00:00:00 GMT+0200', '%a %b %d %Y %H:%M:%S %Z'
 => #<DateTime: 2015-07-06T00:00:00+02:00 ((2457209j,79200s,0n),+7200s,2299161j)>
```

**Comment by marcelog** [comment]
Yup, it's a total mess :\

A PHP example (uses the underlying libc implementation):

```php
var_dump(strptime("Mon Jul 06 2015 00:00:00 ART+0500 (GMT)", "%a %b %d %Y %H:%M:%S %Z%z (%Z)"));
array(9) {
  'tm_sec' =>
  int(0)
  'tm_min' =>
  int(0)
  'tm_hour' =>
  int(16)
  'tm_mday' =>
  int(5)
  'tm_mon' =>
  int(6)
  'tm_year' =>
  int(115)
  'tm_wday' =>
  int(0)
  'tm_yday' =>
  int(185)
  'unparsed' =>
  string(0) ""
}
```

**Comment by bitwalker** [comment]
After thinking about this again, I think I would prefer to leave the current implementation as it is. The last example provided is horribly wrong:

- If I understand the format correctly, the offset is effectively useless, i.e. the timezone of the date/time is ART (which isn't a valid Olson timezone name, but let's ignore that), which should be +5 hours from GMT. However, the effective offset is actually GMT -8 hours.
- ART from what I understand is Argentina time, which is GMT - 3 hours. But GMT-3 + 5 is GMT +2 (if we assume we would allow such arbitrary timezone arithmetic), not GMT-8 as shown in the resulting struct
- `Etc/GMT(-/+)<hours|hours:minutes|hours:minutes:seconds>` is a valid Olson timezone name, but the symbol is actually reversed, i.e. `Etc/GMT-08:00` is actually GMT+8. So combined with the above, we could assume that an offset combined with a timezone name follows this convention, but that makes no sense either, as that convention is specific to the GMT timezone, and does not apply to any other zone.

Basically, as far as I'm concerned, that result is nonsense - it's possible to trace back how the result is achieved, but impossible to predict that result without doing tests. Timex does not allow arbitrary zone arithmetic like this, and thus only supports offsets as part of `GMT` or `Etc/GMT` zone names. This makes it a lot more predictable.

However, what it does seem like we may want here, is a way to say "this format string includes parts which look like an offset, so I want to parse such strings, but ignore this part of the string in the calculation of the resulting date/time". That much does not exist right now, and given the stricter parser, we need such an escape hatch.
